crates/bevy_asset/src/server/mod.rs

```diff
@@ -379,35 +379,48 @@ impl AssetServer {
                 e
             })?;
 
-        let (handle, should_load) = match input_handle {
+        // This contains Some(UntypedHandle), if it was retrievable
+        // If it is None, that is because it was _not_ retrievable, due to
+        //    1. The handle was not already passed in for this path, meaning we can't just use that
+        //    2. The asset has not been loaded yet, meaning there is no existing Handle for it
+        //    3. The path has a label, meaning the AssetLoader's root asset type is not the path's asset type
+        //
+        // In the None case, the only course of action is to wait for the asset to load so we can allocate the
+        // handle for that type.
+        let handle_result = match input_handle {
             Some(handle) => {
                 // if a handle was passed in, the "should load" check was already done
-                (handle, true)
+                Some((handle, true))
             }
             None => {
                 let mut infos = self.data.infos.write();
-                infos.get_or_create_path_handle_untyped(
+                let result = infos.get_or_create_path_handle_internal(
                     path.clone(),
-                    loader.asset_type_id(),
-                    loader.asset_type_name(),
+                    path.label().is_none().then(|| loader.asset_type_id()),
                     HandleLoadingMode::Request,
                     meta_transform,
-                )
+                );
+                unwrap_with_context(result, loader.asset_type_name())
             }
         };
 
-        if path.label().is_none() && handle.type_id() != loader.asset_type_id() {
-            return Err(AssetLoadError::RequestedHandleTypeMismatch {
-                path: path.into_owned(),
-                requested: handle.type_id(),
-                actual_asset_name: loader.asset_type_name(),
-                loader_name: loader.type_name(),
-            });
-        }
-
-        if !should_load && !force {
-            return Ok(handle);
-        }
+        let handle = if let Some((handle, should_load)) = handle_result {
```

Comment:
Realized one hangup with this approach: this means that loading a labeled untyped asset multiple times won't always "early out" for duplicate asset loads. If a second duplicate request for a labeled untyped asset is made before the labeled asset handle is allocated in the AssetLoader, a second load of the root asset will occur.

This isn't broken, but it is suboptimal / undesirable behavior. I'm not sure we can fix this without changing how handles work.

Notably, I think moving to "assets as entities" (or an equivalent "universal id with indirection" system) resolves this problem because you no longer need to know the type to allocate an identifier. This would also allow us to "de-async-ify" untyped asset loading, which is one of the pain points people are experiencing with Bevy Asset V2.

However for a patch release, I think these changes are probably still the move (unless we can find a way to prevent double-loads). I'll give this a bit more thought / if anyone else has ideas let me know.

Reply:
I'm fine to ship with this behavior, especially for a patch release.

Reply:
Shall we open an issue to keep track of this?

Reply:
Opened an issue and added a note to the code!